common/lib/xmodule/xmodule/modulestore/__init__.py

```diff
@@ -139,6 +140,20 @@ def unnest(self):
         """
         self._active_count -= 1
 
+    def add_publish_item(self, location):
+        """
+        Adds an item to the set of things that have been published
+        """
+        self._publish_items.add(location.version_agnostic().for_branch(None))
```

Comment:
Yeah, as long as the re-indexing is happening as a background job, I'd rather stick to doing the stupid-but-simple thing, with some minor optimizations on the tree walking. The `subtree_edited_on` thing would give you a way to cut out big chunks of the tree from your walk, but even then you'd have to worry about deleted nodes and such.